numpy/typing/tests/data/reveal/arithmetic.py

```diff
@@ -0,0 +1,256 @@
+reveal_type(f + c16)  # E: complexfloating
+reveal_type(i + c16)  # E: complexfloating
+reveal_type(AR + c16)  # E: Union[numpy.ndarray, numpy.generic]
+
+reveal_type(c8 + c16)  # E: complexfloating
+reveal_type(c8 + f8)  # E: complexfloating
+reveal_type(c8 + i8)  # E: complexfloating
+reveal_type(c8 + c8)  # E: complexfloating
+reveal_type(c8 + f4)  # E: complexfloating
+reveal_type(c8 + i4)  # E: complexfloating
+reveal_type(c8 + b_)  # E: complex64
+reveal_type(c8 + b)  # E: complexfloating
+reveal_type(c8 + c)  # E: complexfloating
+reveal_type(c8 + f)  # E: complexfloating
+reveal_type(c8 + i)  # E: complexfloating
+reveal_type(c8 + AR)  # E: Union[numpy.ndarray, numpy.generic]
+
+reveal_type(c16 + c8)  # E: complexfloating
+reveal_type(f8 + c8)  # E: complexfloating
+reveal_type(i8 + c8)  # E: complexfloating
+reveal_type(c8 + c8)  # E: complexfloating
+reveal_type(f4 + c8)  # E: complexfloating
+reveal_type(i4 + c8)  # E: complexfloating
+reveal_type(b_ + c8)  # E: complex64
+reveal_type(b + c8)  # E: complexfloating
+reveal_type(c + c8)  # E: complexfloating
+reveal_type(f + c8)  # E: complexfloating
+reveal_type(i + c8)  # E: complexfloating
+reveal_type(AR + c8)  # E: Union[numpy.ndarray, numpy.generic]
+
+# Float
+
+reveal_type(f8 + f8)  # E: floating
+reveal_type(f8 + i8)  # E: floating
+reveal_type(f8 + f4)  # E: floating
+reveal_type(f8 + i4)  # E: floating
+reveal_type(f8 + b_)  # E: float64
+reveal_type(f8 + b)  # E: floating
+reveal_type(f8 + c)  # E: complexfloating
+reveal_type(f8 + f)  # E: floating
+reveal_type(f8 + i)  # E: floating
+reveal_type(f8 + AR)  # E: Union[numpy.ndarray, numpy.generic]
+
+reveal_type(f8 + f8)  # E: floating
+reveal_type(i8 + f8)  # E: floating
+reveal_type(f4 + f8)  # E: floating
+reveal_type(i4 + f8)  # E: floating
+reveal_type(b_ + f8)  # E: float64
+reveal_type(b + f8)  # E: floating
+reveal_type(c + f8)  # E: complexfloating
+reveal_type(f + f8)  # E: floating
+reveal_type(i + f8)  # E: floating
+reveal_type(AR + f8)  # E: Union[numpy.ndarray, numpy.generic]
+
+reveal_type(f4 + f8)  # E: floating
+reveal_type(f4 + i8)  # E: floating
+reveal_type(f4 + f4)  # E: floating
+reveal_type(f4 + i4)  # E: floating
+reveal_type(f4 + b_)  # E: float32
+reveal_type(f4 + b)  # E: floating
+reveal_type(f4 + c)  # E: complexfloating
+reveal_type(f4 + f)  # E: floating
+reveal_type(f4 + i)  # E: floating
+reveal_type(f4 + AR)  # E: Union[numpy.ndarray, numpy.generic]
+
+reveal_type(f8 + f4)  # E: floating
+reveal_type(i8 + f4)  # E: floating
+reveal_type(f4 + f4)  # E: floating
+reveal_type(i4 + f4)  # E: floating
+reveal_type(b_ + f4)  # E: float32
+reveal_type(b + f4)  # E: floating
+reveal_type(c + f4)  # E: complexfloating
+reveal_type(f + f4)  # E: floating
+reveal_type(i + f4)  # E: floating
+reveal_type(AR + f4)  # E: Union[numpy.ndarray, numpy.generic]
+
+# Int
+
+reveal_type(i8 + i8)  # E: signedinteger
+reveal_type(i8 + u8)  # E: Union[numpy.signedinteger, numpy.floating]
```

Comment:
Note that the output I'd expect is
```python
reveal_type(i8 + u8)  # E: numpy.int16
```
Fine to dismiss that as future work, but a comment next to `# Int` or at the top of the file saying that many of these patterns are overly vague would be nice.

Reply:
The `floating` case is specifically for `uint64 + signedinteger -> float64`, though I suppose this does mean that `floating` can be replaced with `float64`.

https://github.com/numpy/numpy/pull/17273/commits/9e63ecf75d6055c3f8b424e7badb28773455b082 contains a few additional notes on the subject in addition to the `floating` / `float64` swap.

Reply:
And yes, further refining of the return types (_e.g._ `int64 + float32 -> float64`) is planned for a future PR.
Just have to think a bit first about be the best way to implement this.